arcgis-ios-sdk-samples/Geometry/List transformations/ListTransformationsViewController.swift

```diff
@@ -0,0 +1,178 @@
+            let peDataURL = URL(fileURLWithPath: path[0]).appendingPathComponent(documentPEDataRuntime)
+            try? AGSTransformationCatalog.setProjectionEngineDirectory(peDataURL)
+        }
+        
+        setupTransformsList()
+    }
+
+    @IBAction func oderByMapExtentValueChanged(_ sender: Any) {
+        setupTransformsList()
+    }
+
+    //MARK: - TableView data source
+
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return datatumTransformations.count
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: "DatumTransformCell", for: indexPath)
+
+        // get the selected transformation
+        let transformation = datatumTransformations[indexPath.row]
+
+        // if we're missing the grid files, detail which ones
+        var files = ""
+        if transformation.isMissingProjectionEngineFiles {
+            files = "Missing grid files"
+
+            if let gt = transformation as? AGSGeographicTransformation {
+                gt.steps.forEach { (step) in
+                    if step.isMissingProjectionEngineFiles {
+                        files.append(": " + step.projectionEngineFilenames.joined(separator: ","))
+                    }
+                }
+            }
+        }
+        
+        cell.textLabel?.text = transformation.name
+        cell.detailTextLabel?.text = files
+        
+        if let defaultTransform = defaultTransformation {
+            cell.isSelected = transformation.isEqual(to: defaultTransform)
+        }
+        else {
+            cell.isSelected = false
+        }
+        
+        return cell
+    }
+    
+    //MARK: - TableView delegates
+    
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        guard let mapViewSR = mapView.spatialReference else { return }
+
+        let selectedTransform = datatumTransformations[indexPath.row]
+        if let projectedGeometry = AGSGeometryEngine.projectGeometry(originalGeometry, to: mapViewSR, datumTransformation: selectedTransform) {
+            // projectGeometry succeeded
+            if graphicsOverlay.graphics.count > 1,
+                let graphics = graphicsOverlay.graphics as? [AGSGraphic],
+                let graphic = (graphics).last {
+                // we've already added the projected graphic
+                graphic.geometry = projectedGeometry
+            }
+            else {
+                // add projected graphic
+                addGraphic(projectedGeometry, color: .blue, style: .cross)
+            }
+        }
+        else {
+            // If a transformation is missing grid files, then it cannot be
+            // successfully used to project a geometry, and "projectGeometry" will return nil.
+            // In that case, remove projected graphic
+            if graphicsOverlay.graphics.count > 1 {
```

Comment:
It might be better to prevent the user from selecting invalid transformations in the first place. One way of doing that is to disable user interaction on those cells in `tableView(_:cellForRowAt:)`:

```swift
cell.isUserInteractionEnabled = !transformation.isMissingProjectionEngineFiles
```

Reply:
Good call.